File: problems/HackerRank/sortedInsert.ts
import { DoublyLinkedListNode } from './reverseDoublyLinkedList';

function sortedInsert(
  llist: DoublyLinkedListNode,
  data: number
): DoublyLinkedListNode {
  let list = llist;
  let cur = llist;

  while (!!cur) {
    // data is in the head
    if (data <= list.data) {
      const node = new DoublyLinkedListNode(data);
      node.next = list;
      list.prev = node;
      list = list.prev;
      break;
    }

    // data is in the middle
    if (data >= cur.data && cur?.next && data < cur?.next?.data) {
      const node = new DoublyLinkedListNode(data);
      node.next = cur.next;
      node.prev = cur;
      cur.next.prev = node;
      cur.next = node;
      break;
    }

    // data is in the end
    if (data >= cur.data && !cur?.next) {
      const node = new DoublyLinkedListNode(data);
      node.prev = cur;
      cur.next = node;
      break;
    }

    cur = cur.next as DoublyLinkedListNode;
  }

  return list;
}

File: problems/HackerRank/2521_exercises/linkedList/easy/sortedInsert.c
DoublyLinkedListNode *sortedInsert(DoublyLinkedListNode *llist, int data)
{

	struct DoublyLinkedListNode *newNode = (struct DoublyLinkedListNode *)malloc(sizeof(DoublyLinkedListNode));
	newNode->data = data;
	newNode->next = NULL;
	newNode->prev = NULL;

	// if the list is empty, return the newNode as the head
	if (llist == NULL)
	{
		return newNode;
	}

	// if the newNode has to insert at the beginning
	if (llist->data > data)
	{
		newNode->next = llist;
		llist->prev = newNode;
		return newNode;
	}

	//  1,3,4,5,10
	// traverse the list to find the correct postion
	struct DoublyLinkedListNode *current = llist;
	while (current->next != NULL && current->next->data < data)
	{
		current = current->next;
	}

	// insert the new node at the specified postion
	newNode->prev = current;
	newNode->next = current->next;
	if (current->next != NULL)
	{
		current->next->prev = newNode;
	}
	current->next = newNode;

	return llist;
}
